File: src/main/java/tn/esprit/entities/Commentaires.java
package tn.esprit.entities;

import java.util.Date;

public class Commentaires {

    // ── Enum status — valeurs réelles en BD : ACTIF, SUPPRIME, SIGNALE ───────
    public enum StatusCommentaires {
        ACTIF, SUPPRIME, SIGNALE, ARCHIVE
    }

    // ── Champs — EXACTEMENT les colonnes de la table `commentaire` ────────────
    private int                id_commentaires;  // PK AUTO_INCREMENT
    private int                id_user;          // INT  → colonne id_user
    private String             user_name;        // VARCHAR(255)
    private String             contenu;          // LONGTEXT
    private Date               datePublication;  // DATE → colonne date_publication
    private int                likes;            // INT DEFAULT 0
    private int                reponse;          // INT DEFAULT 0
    private StatusCommentaires status;           // VARCHAR(20) DEFAULT 'ACTIF'
    private String             sentiment;        // VARCHAR(20) nullable
    private String             themes;           // LONGTEXT JSON nullable
    private String             aiSummary;        // LONGTEXT → colonne ai_summary nullable
    private int                ressource_id;     // INT → colonne ressource_id (FK)
    private Integer            parent_id;        // INT nullable → réponse imbriquée

    // ── Constructeur vide ─────────────────────────────────────────────────────
    public Commentaires() {}

    // ── Constructeur complet (colonnes BD) ────────────────────────────────────
    public Commentaires(int id_commentaires, int id_user, String user_name,
                        String contenu, Date datePublication, int likes,
                        int reponse, StatusCommentaires status,
                        String sentiment, String themes, String aiSummary,
                        int ressource_id, Integer parent_id) {
        this.id_commentaires = id_commentaires;
        this.id_user         = id_user;
        this.user_name       = user_name;
        this.contenu         = contenu;
        this.datePublication = datePublication;
        this.likes           = likes;
        this.reponse         = reponse;
        this.status          = status;
        this.sentiment       = sentiment;
        this.themes          = themes;
        this.aiSummary       = aiSummary;
        this.ressource_id    = ressource_id;
        this.parent_id       = parent_id;
    }

    // ── Constructeur simplifié (ajout rapide) ─────────────────────────────────
    public Commentaires(int ressource_id, int id_user, String user_name, String contenu) {
        this.ressource_id    = ressource_id;
        this.id_user         = id_user;
        this.user_name       = user_name;
        this.contenu         = contenu;
        this.status          = StatusCommentaires.ACTIF;
        this.likes           = 0;
        this.reponse         = 0;
        this.datePublication = new java.util.Date();
    }

    // ── Getters / Setters ─────────────────────────────────────────────────────
    public int    getId_commentaires()            { return id_commentaires; }
    public void   setId_commentaires(int v)       { this.id_commentaires = v; }

    public int    getId_user()                    { return id_user; }
    public void   setId_user(int v)               { this.id_user = v; }

    /** Alias rétrocompat */
    public int    getId_User()                    { return id_user; }
    public void   setId_User(int v)               { this.id_user = v; }

    /** Alias rétrocompat : getId_Article() → ressource_id */
    public int    getId_Article()                 { return ressource_id; }
    public void   setId_Article(int v)            { this.ressource_id = v; }

    public int    getRessource_id()               { return ressource_id; }
    public void   setRessource_id(int v)          { this.ressource_id = v; }

    public String getUser_name()                  { return user_name; }
    public void   setUser_name(String v)          { this.user_name = v; }

    public String getContenu()                    { return contenu; }
    public void   setContenu(String v)            { this.contenu = v; }

    public Date   getDate()                       { return datePublication; }
    public void   setDate(Date v)                 { this.datePublication = v; }

    public Date   getDatePublication()            { return datePublication; }
    public void   setDatePublication(Date v)      { this.datePublication = v; }

    public int    getLikes()                      { return likes; }
    public void   setLikes(int v)                 { this.likes = v; }

    public int    getReponse()                    { return reponse; }
    public void   setReponse(int v)               { this.reponse = v; }

    public StatusCommentaires getStatus()         { return status; }
    public void   setStatus(StatusCommentaires v) { this.status = v; }

    public void setStatusFromString(String s) {
        if (s == null) { this.status = StatusCommentaires.ACTIF; return; }
        try { this.status = StatusCommentaires.valueOf(s.toUpperCase()); }
        catch (IllegalArgumentException e) { this.status = StatusCommentaires.ACTIF; }
    }

    public boolean isSignale() { return status == StatusCommentaires.SIGNALE; }

    public String getSentiment()                  { return sentiment; }
    public void   setSentiment(String v)          { this.sentiment = v; }

    public String getThemes()                     { return themes; }
    public void   setThemes(String v)             { this.themes = v; }

    public String getAiSummary()                  { return aiSummary; }
    public void   setAiSummary(String v)          { this.aiSummary = v; }

    public Integer getParent_id()                 { return parent_id; }
    public void    setParent_id(Integer v)        { this.parent_id = v; }

    @Override
    public String toString() {
        return "Commentaires{" +
                "id=" + id_commentaires +
                ", id_user=" + id_user +
                ", user_name='" + user_name + '\'' +
                ", contenu='" + contenu + '\'' +
                ", date=" + datePublication +
                ", likes=" + likes +
                ", status=" + status +
                ", ressource_id=" + ressource_id +
                ", parent_id=" + parent_id +
                '}';
    }
}

File: src/Entity/Commentaires.php
<?php
namespace App\Entity;

use App\Repository\CommentairesRepository;
use App\Validator\NoBadWords;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity(repositoryClass: CommentairesRepository::class)]
#[ORM\Table(name: 'commentaire')]
class Commentaires
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'id_commentaires')]
    /** @phpstan-ignore-next-line */
    private ?int $id = null;

   #[ORM\ManyToOne(inversedBy: 'commentaires')]
#[ORM\JoinColumn(name: 'ressource_id', referencedColumnName: 'id_ressources', onDelete: 'CASCADE', nullable: false)]
private ?Ressources $ressource = null;

  #[ORM\Column(name: 'id_user', type: 'integer')]
private int $idUser = 0;



    #[ORM\Column(name: 'user_name', length: 255)]
    #[Assert\NotBlank(message: "Le nom d'utilisateur est obligatoire.")]
    private string $user_name = '';

    #[ORM\Column(type: Types::TEXT)]
    #[Assert\NotBlank(message: 'Le contenu du commentaire est obligatoire.')]
    #[Assert\Length(max: 5000, maxMessage: 'Le commentaire ne doit pas dépasser 5000 caractères.')]
    #[NoBadWords]
    private string $contenu = '';

    #[ORM\Column(name: 'date_publication', type: Types::DATE_MUTABLE)]
    private \DateTimeInterface $datePublication;

    #[ORM\Column]
    private int $likes = 0;

    #[ORM\Column]
    private int $reponse = 0;

    #[ORM\Column(type: 'string', length: 20, options: ['default' => 'ACTIF'])]
    #[Assert\Choice(choices: ['ACTIF', 'SUPPRIME', 'SIGNALE'], message: 'Statut invalide.')]
    private string $status = 'ACTIF';

    #[ORM\Column(length: 20, nullable: true)]
    private ?string $sentiment = null;

    /**
     * @var array<int, string>|null
     */
    #[ORM\Column(type: 'json', nullable: true)]
    private ?array $themes = null;

    #[ORM\Column(name: 'ai_summary', type: Types::TEXT, nullable: true)]
    private ?string $aiSummary = null;

    #[ORM\ManyToOne(targetEntity: self::class, inversedBy: 'reponses')]
    #[ORM\JoinColumn(name: 'parent_id', referencedColumnName: 'id_commentaires', nullable: true, onDelete: 'CASCADE')]
    private ?self $parent = null;

    /**
     * @var Collection<int, Commentaires>
     */
    #[ORM\OneToMany(mappedBy: 'parent', targetEntity: self::class, cascade: ['remove'])]
    private Collection $reponses;

    public function __construct()
    {
        $this->reponses        = new ArrayCollection();
        $this->datePublication = new \DateTime();
    }

    public function getId(): ?int { return $this->id; }

    public function getRessource(): ?Ressources { return $this->ressource; }
    public function setRessource(?Ressources $ressource): static { $this->ressource = $ressource; return $this; }

    public function getIdUser(): int { return $this->idUser; }
    public function setIdUser(int $idUser): static { $this->idUser = $idUser; return $this; }

    public function getUserName(): string { return $this->user_name; }
    public function setUserName(string $user_name): static { $this->user_name = $user_name; return $this; }

    public function getContenu(): string { return $this->contenu; }
    public function setContenu(string $contenu): static { $this->contenu = $contenu; return $this; }

    public function getDatePublication(): \DateTimeInterface { return $this->datePublication; }
    public function setDatePublication(\DateTimeInterface $datePublication): static { $this->datePublication = $datePublication; return $this; }

    public function getLikes(): int { return $this->likes; }
    public function setLikes(int $likes): static { $this->likes = $likes; return $this; }
    public function incrementLikes(): static { $this->likes++; return $this; }

    public function getReponse(): int { return $this->reponse; }
    public function setReponse(int $reponse): static { $this->reponse = $reponse; return $this; }

    public function getStatus(): string { return $this->status; }
    public function setStatus(string $status): static { $this->status = $status; return $this; }

    public function getSentiment(): ?string { return $this->sentiment; }
    public function setSentiment(?string $sentiment): static { $this->sentiment = $sentiment; return $this; }

    /**
     * @return array<int, string>|null
     */
    public function getThemes(): ?array { return $this->themes; }

    /**
     * @param array<int, string>|null $themes
     */
    public function setThemes(?array $themes): static { $this->themes = $themes; return $this; }

    public function getAiSummary(): ?string { return $this->aiSummary; }
    public function setAiSummary(?string $aiSummary): static { $this->aiSummary = $aiSummary; return $this; }

    public function getParent(): ?self { return $this->parent; }
    public function setParent(?self $parent): static { $this->parent = $parent; return $this; }

    /**
     * @return Collection<int, Commentaires>
     */
    public function getReponses(): Collection { return $this->reponses; }

    public function addReponse(self $reponse): static
    {
        if (!$this->reponses->contains($reponse)) {
            $this->reponses->add($reponse);
            $reponse->setParent($this);
        }
        return $this;
    }

    public function removeReponse(self $reponse): static
    {
        if ($this->reponses->removeElement($reponse)) {
            if ($reponse->getParent() === $this) {
                $reponse->setParent(null);
            }
        }
        return $this;
    }
}